[RFC] Improve tests

Shougo opened the pull request:
https://github.com/neovim/neovim/pull/14534

I have fixed the tests in the first.

Shougo committed:
vim-patch:8.2.3616: arglist test does not clear the argument list consistently

Problem:    Arglist test does not clear the argument list consistently.
Solution:   Call Reset_arglist(). (Shougo Matsushita, closes vim/vim#9154)
https://github.com/vim/vim/commit/3cad47038547e694cfa26ba39c399f610d2054bd

Shougo commented:
I think it is not related my changes.

```
stack traceback:
	test\helpers.lua:73: in function 'eq'
	test/functional\core\job_spec.lua:359: in function <test/functional\core\job_spec.lua:353>


 53 SKIPPED TESTS
 1 ERROR
-- Tests exited non-zero: 1
-- Output to stderr:
```

Shougo commented:
@janlazo Can you check it?

jamessan commented:
What's the reasoning behind these test changes?  The tests are running cleanly already, so why do we need to make changes that diverge from Vim?

Shougo commented:
> What's the reasoning behind these test changes? The tests are running cleanly already, so why do we need to make changes that diverge from Vim?

OK.  I have added some `cd` to test it locally.
If cwd is not home, it seems failed.  I think the tests should be cwd independent.

And I have added some `delete()` to test it locally.
If you execute it locally, previous test files my exists.
I think `delete()` is better.

Shougo commented:
So current tests are tested/created for only CI or cleaned environment.
I think tests can be executed locally to tests patches.

Am I mistake??

janlazo commented:
This means that there are unmerged vim patches to delete old test data.

Shougo commented:
> This means that there are unmerged vim patches to delete old test data.

No.  You can check Vim's source code.
My test fixes are not patch porting.

janlazo commented on src/nvim/testdir/test_registers.vim:
```diff
@@ -43,6 +43,9 @@ func Test_yank_shows_register()
 
 func Test_display_registers()
+    " Disable clipboard
+    let g:clipboard = {}
```
This is fine but put in separate commit.

Shougo replied:
Done.

janlazo replied:
Forgot to ask.
Does this affect the other tests in this file in case that this fails?

Shougo replied:
Please see it.

https://github.com/neovim/neovim/pull/12784#issuecomment-683495832

It is neovim only problem.

janlazo replied:
Without your changes, I can reproduce that error on my machine depending on the clipboard state.

What I'm asking is if it impacts the other tests in case `Test_display_registers()` because `g:clipboard` is not restored. Vim tests typically restore the previous state of the test runner within the test such that the test must pass to not break the next tests.
It's fine if you don't know.

Shougo replied:
OK.  I have added the restore.

Shougo committed:
fix: add cd for local execution

Shougo committed:
fix: remove previous executed directories to execute tests locally

Shougo committed:
fix: disable clipboard when test registers

Shougo commented:
@janlazo Ping.  Can I port the patch?

janlazo commented:
Yes.

janlazo commented on src/nvim/testdir/test_quickfix.vim:
```diff
@@ -1012,9 +1012,9 @@ endfunc
-  call mkdir('dir1/c')
+  call mkdir('dir1/a', 'p')
+  call mkdir('dir1/a/b', 'p')
+  call mkdir('dir1/c', 'p')
```
Not seeing the value of the `'p'` flag when the previous `mkdir()` aren't collapsed to 1 line.
This isn't in Vim but it can be upstreamed.

janlazo commented on src/nvim/testdir/test_command_count.vim:
```diff
@@ -103,6 +103,8 @@ endfunc
 
 func Test_command_count_2()
   silent! %argd
+  cd
```
Vim doesn't have this.
`:cd` changes to the home directory in Unix system.
Why?

Shougo replied:
I will remove it.

janlazo commented on src/nvim/testdir/test_arglist.vim:
```diff
@@ -1,5 +1,10 @@
 " Test argument list commands
 
+func Reset_arglist()
+  cd
```
https://github.com/neovim/neovim/pull/14540#discussion_r636577072
Same comment about `cd` for Unix systems.

janlazo commented:
@Shougo It's beneficial for both projects for you to upstream your improvements to the tests. Don't use `:cd` without args because the documented behavior suggests that it's not portable on non-Unix systems (unless the doc is wrong now).

Shougo committed:
fix: for reviews

Shougo committed:
fix: remove unnecessary changes

Shougo commented:
Hm...


`let $GZIP = ""` does not work for me.
`[,']!GZIP= gzip -d` works instead.

Shougo committed:
fix: fix GZIP error

Shougo commented:
Almost tests are passed.

janlazo commented on test/functional/legacy/011_autocommands_spec.lua:
```diff
@@ -67,10 +67,15 @@ describe('file reading, writing and bufnew and filter autocommands', function()
       feed_command('set bin')
-      feed_command("au FileReadPost    *.gz   '[,']!gzip -d")
+      if iswin() then
+          feed_command('let $GZIP = ""')
```
`call setenv('GZIP', v:null)` for both.

Shougo replied:
Unfortunately, it does not work for me.
So I cannot accept the suggestion.

```
stack traceback:
	test/helpers.lua:73: in function 'expect'
	test/functional/legacy/011_autocommands_spec.lua:76: in function <test/functional/legacy/011_autocommands_spec.lua:68>


 1 ERROR
-- Tests exited non-zero: 1
-- Output to stderr:
gzip: warning: GZIP environment variable is deprecated; use an alias or script
Xtestfile:	 73.9% -- replaced with Xtestfile.gz
gzip: warning: GZIP environment variable is deprecated; use an alias or script
Xtestfile:	 73.9% -- replaced with Xtestfile.gz
gzip: warning: GZIP environment variable is deprecated; use an alias or script
Xtestfile:	 73.9% -- replaced with Xtestfile.gz
```

janlazo replied:
Does your shell set `GZIP`? Does the problem remain if you change the shell to `sh`?

Shougo replied:
OH, it fixes my problem!
Thanks.

Shougo committed:
fix: shell problem

Shougo commented:
I have pushed the fix.

Shougo committed:
test: save clipboard

Shougo committed:
fix: error

Shougo committed:
fix: remove unneeded gzip check

Shougo commented:
Can you review it again?

Shougo commented:
Almost succeeded.

Shougo commented:
It should be only change tests instead.

Shougo commented:
I have rebased tests.  Please review it again.

Shougo commented:
Some tests are failed in Windows CI.
But they are same with master.  My changes are not affected.

Shougo commented:
Can you review it and merge?

janlazo commented on src/nvim/testdir/test_arglist.vim:
```diff
@@ -26,6 +30,8 @@ func Test_argidx()
 endfunc
 
 func Test_argadd()
+  call Reset_arglist()
```
I don't see the rationale in the commit message and it's not in Vim.
What's the issue?

janlazo replied:
I suppose that this is to match `Test_zero_argadd`.

Shougo replied:
```
  From test_command_count.vim:
  Found errors in Test_command_count_2():
  function RunTheTest[37]..Test_command_count_2 line 9: Expected 'a' but got ''
```

I think arglist reset is needed.

janlazo commented on src/nvim/testdir/test_arglist.vim:
```diff
@@ -115,8 +121,7 @@ endfunc
 func Test_argument()
-  " Clean the argument list
-  arga a | %argd
+  call Reset_arglist()
```
Can you send this to Vim?

Shougo replied:
Done.

janlazo commented:
My only issue is the `Reset_arglist()`  changes that Vim does not have in some tests. I'm fine with the rest.

Shougo commented:
`Reset_arglist()` seems not needed.

Shougo commented:
Done.

https://github.com/vim/vim/pull/9154

Shougo commented:
Merged in https://github.com/vim/vim/commit/3cad47038547e694cfa26ba39c399f610d2054bd

Shougo commented:
@janlazo I think it has no problem.

janlazo commented:
Redo one of the commits as a vim-patch and I'll merge.

Shougo commented:
I have added vim-patch 8.2.3616 commit.

janlazo commented:
Thank you for your patience and cooperation.

Shougo commented:
Thank you!
